stdlib/string.gr

```diff
@@ -1609,3 +1609,66 @@ export let rec forEachCodePointi = (fn: (Number, Number) -> Void, str: String) =
   Memory.decRef(WasmI32.fromGrain(forEachCodePointi))
   void
 }
+
+/**
+ * Trims a string removing any leading or following white spaces based on the end param.
+ *
+ * @param str: The string to be trimmed.
+ * @param end: If true we are trimming the following spaces, otherwise trim leading spaces.
+ * @returns The trimmed string.
+ * 
+ * @example String.trimString("   Hello World   ", false)
+ */
+let trimString = (str: String, end: Bool) => {
+  let chars = explode(str), charsLength = length(str)
+  let mut i = 0, offset = 1
+  if (end) {
+    i = charsLength-1
+    offset = -1
+  }
+  for (; i < charsLength && i > -1; i += offset) {
+    let currentChar = chars[i];
+    // TODO: Use unicode whitespace property and unicode line terminator once github issue #661 is completed
+    if (
+      // Spacing
+      currentChar != '\u{0009}'  && // Tab
+      currentChar != '\u{000B}'  && // LINE TABULATION
+      currentChar != '\u{000C}'  && // FORM FEED (FF)
+      currentChar != '\u{0020}'  && // Space
+      currentChar != '\u{00A0}'  && // No Break Space
+      currentChar != '\u{FEFF}' && // 	ZERO WIDTH NO-BREAK SPACE
+      // Line Terminator
+      currentChar != '\n' && // LF
+      currentChar != '\r' // CR
+    ) break
+  }
+  if (end) slice(0, i+1, str)
+  else slice(i, charsLength, str)
+}
+/**
+ * Trims the `start` of a string removing any leading white spaces.
+ *
+ * @param str: The string to be trimmed.
+ * @returns The trimmed string.
+ * 
+ * @example String.trimStart("   Hello World")
+ */
+export let trimStart = (str: String) => trimString(str, false)
+/**
+ * Trims the `end` of a string removing any following white spaces.
+ *
+ * @param str: The string to be trimmed.
+ * @returns The trimmed string.
+ * 
+ * @example String.trimEnd("Hello World   ")
+ */
+export let trimEnd = (str: String) => trimString(str, true)
+/**
+ * Trims a string removing leading and following white spaces.
+ *
+ * @param str: The string to be trimmed.
+ * @returns The trimmed string.
+ * 
+ * @example String.trim("   Hello World   ")
+ */
+export let trim = (str: String) => trimEnd(trimStart(str))
```

Comment:
An additional question on the implementation: do we always want to call 2 functions to trim a string from both ends? Is there a more efficient way to do it?

Reply:
I could use an enum as the input for the internal function with Start, End, Both as values and then just only have one function call. I cant think of and couldn't find any faster implementation but if you know of one I would love to take a crack at implementing it.

Reply:
I wasn't thinking about an enum, but more concerned that you are doing 2 loops on the string. Let me poke around a bit.

Reply:
That's what I figured, I could possibly do it in one loop, and loop through both sides at once and then just break on the longer side marking the lengths of each side, I figure that would not increase speed by that much though anyways and may make the code a little a messy.

Reply:
It seems that Ruby implements it as 2 separate loops (actually a **lot** of loops, but let's call it 2) while lodash relies on using RegExps. I think I'm fine with the 2-loop approach here unless @peblair or @ospencer has a better solution.

Reply:
I dont know how much aot optimization the compiler does on regex i assumed it wasnt much and figured the loop would be more efficent. if the regex method is more efficent i would be happy implementing that.

Reply:
@spotandjake I don't believe that a Regex would be more efficient than the current implementation (and would add a large dependency to the String stdlib module), so we probably don't want to use it.

Reply:
My thought was to add 2 loops inside your helper function depending on the start/end flags, which would avoid an extra `explode` and `length` call.

Reply:
To optimize allocations for large strings it may make sense to use .charAt over chars[i] with the String.explode if the string is over a certain length. this way if people have massive data a big array isnt being allocated and if they have smaller data they are getting the fastest method available.

Reply:
The only problem I see with `charAt` is that it is `O(n)` because Grain strings are UTF-8 (meaning that the worst-case complexity of this function (called with a string of spaces) would be `O(n^2)`). I would agree that some sort of heuristic based on the length to decide whether to explode would make sense. Additionally, I might throw in some sort of basic `if (!isWhitespace(String.charAt(0))) {...}` / `if (!isWhitespace(String.charAt(-1))) {...}` check at the beginning to bail out quickly. Finally, I'm +1 on the two-loop solution.